rust/operator-binary/src/pod_svc_controller.rs

```diff
@@ -47,21 +58,28 @@ pub struct Ctx {
 
 const APP_KUBERNETES_LABEL_BASE: &str = "app.kubernetes.io/";
 
-pub async fn reconcile_pod(pod: Arc<Pod>, ctx: Arc<Ctx>) -> Result<Action, Error> {
+pub async fn reconcile_pod(pod: Arc<Pod>, ctx: Arc<Ctx>) -> Result<Action> {
     tracing::info!("Starting reconcile");
 
     let name = pod.metadata.name.clone().context(PodHasNoNameSnafu)?;
 
-    let pod_labels = pod
-        .metadata
-        .labels
-        .as_ref()
-        .with_context(|| PodHasNoLabelsSnafu { name: name.clone() })?;
+    let pod_labels = Labels::try_from(
+        pod.metadata
+            .labels
+            .as_ref()
+            .with_context(|| PodHasNoLabelsSnafu { name: name.clone() })?,
+    )
+    .context(BuildLabelSnafu)?;
 
     let recommended_labels_from_pod = pod_labels
         .iter()
-        .filter(|(key, _)| key.starts_with(APP_KUBERNETES_LABEL_BASE))
-        .map(|(key, value)| (key.clone(), value.clone()))
+        .filter(|&label| {
+            label
+                .key()
+                .prefix()
+                .is_some_and(|prefix| *prefix == APP_KUBERNETES_LABEL_BASE)
```

Comment:
```suggestion
                == Some(APP_KUBERNETES_LABEL_BASE)
```

Reply:
Unfortunately not:

```
mismatched types
expected reference `&KeyPrefix`
   found reference `&'static str`rustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message%20%5B0%5D?0#file%3A%2F%2F%2Fhome%2Fnick%2Fdev%2Fgithub.com%2Fstackabletech%2Fhdfs-operator%2Frust%2Foperator-binary%2Fsrc%2Fpod_svc_controller.rs)
```

And using `KeyPrefix::from_str(...)` or ``KeyPrefix::try_from(...)` causes a result.

From memory I had already tried doing something like `impl<T> PartialEq<&str> for Option<&KeyPrefix>`, but of course, The trait and the enum are foreign.

Any suggestions, or leave as is?

Reply:
`KeyPrefix` is `Deref<Target = String>`, which is in turn `Deref<Target = str>`, so try `.prefix().as_deref().as_deref() == ...`. `&String` is kind of useless, ideally `KeyPrefix` would be changed to be `Deref<Target = str>` instead (which would let us get rid of one `.as_deref()`).

Reply:
Hmm, none of these are working:

```rust
        .filter(|&label| {
            label.key().prefix().as_deref().as_deref() == Some(APP_KUBERNETES_LABEL_BASE)
        })
```

```
mismatched types
expected reference `&KeyPrefix`
   found reference `&'static str`rustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message%20%5B0%5D?0#file%3A%2F%2F%2Fhome%2Fnick%2Fdev%2Fgithub.com%2Fstackabletech%2Fhdfs-operator%2Frust%2Foperator-binary%2Fsrc%2Fpod_svc_controller.rs)
```


But honestly, I don't think that is much tidier than how it was:

```rust
                .is_some_and(|prefix| *prefix == APP_KUBERNETES_LABEL_BASE)
```

Reply:
Aight, fair.

Reply:
@Techassi is looking into the `Deref<Target = str>`, so could be something to come back to

Reply:
I implemented the change from `String` to `str` as the `Deref` target in https://github.com/stackabletech/operator-rs/pull/725. While adding a few tests for this feature, I encountered a rather peculiar issue:

```rust
let label = Label::try_from((key, "zookeeper")).unwrap();

// Works as expected
let is_valid = label
    .key()
    .prefix()
    .is_some_and(|prefix| *prefix == "app.kubernetes.io");

// Doesn't work as expected
let is_valid = label.key().prefix().as_deref() == Some("app.kubernetes.io");
```

Using `as_deref()` on an `Option<&KeyPrefix>` will return `&KeyPrefix` instead of `&str`. A simplified code snippet demonstrates this rather nicely ([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=57010767763cf731d523ef0b9809231e)).

Manually implementing `Deref for &T` collides with a blanket implemention in `core`. Any idea what we can do in this case?

Reply:
Oh cool, yeah this works (double deref until the next operator-rs release):

```rust
            label
                .key()
                .prefix()
                .map(std::ops::Deref::deref)
                .map(std::ops::Deref::deref)
                == Some(APP_KUBERNETES_LABEL_BASE)
```

> if you prefer sigil soup

I've already eaten, thanks :P